rfcs/text/0013_background_sessions.md

```diff
@@ -0,0 +1,503 @@
+    sessionId: string
+  ) => Promise<string?>
+
+   /**
+    * Store a session.
+    * @param request 
+    * @param sessionId Session ID to store. Probably retrieved from `sessionService.get()`.
+    * @param searchIdMap A mapping of hashed requests mapped to the corresponding searchId.
+    * @param url TODO: is the URL provided here? How?
+    * @returns The stored `BackgroundSessionAttributes` object
+    * @throws Throws an error in OSS.
+    * @internal (Consumers should use searchInterceptor.sendToBackground())
+    */
+  store: (
+    request: KibanaRequest,
+    sessionId: string,
+    name: string,
+    url: string,
+    searchIdMapping?: Record<string, string>
+  ) => Promise<BackgroundSessionAttributes>
+
+   /**
+    * Mark a session as and all associated searchIds as expired.
+    * @param request 
+    * @param sessionId
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+  expire: async (
+    request: KibanaRequest,
+    sessionId: string
+  ) => Promise<boolean>
+
+   /**
+    * Extend a session and all associated searchIds.
+    * @param request 
+    * @param sessionId 
+    * @param extendBy Time to extend by, can be a relative or absolute string.
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+  extend: async (
+    request: KibanaRequest,
+    sessionId: string,
+    extendBy: string,
+  ) => Promise<boolean>
+
+   /**
+    * Get a list of background session objects.
+    * @param request 
+    * @param sessionId
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+  list: async (
+    request: KibanaRequest,
+  ) => Promise<BackgroundSessionAttributes[]>
+
+   /**
+    * Update the status of a given session
+    * @param request 
+    * @param sessionId
+    * @param status
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+  updateStatus: async (
+    request: KibanaRequest,
+    sessionId: string,
+    status: BackgroundSessionStatus
+  ) => Promise<boolean>
+}
+
+```
+
+### Search Service Changes
+
+There are cases where search requests are issued by the server (Like TSVB).
+We can simplify this flow by introducing a mechanism, similar to the frontend one, tracking the information in memory and polling for a saved object with a corresponding sessionId to store the ids into it.
+
+```ts
+interface SearchService {
+  /**
+   * The search API will accept the option `trackId`, which will track the search ID, if available, on the server, until a corresponding saved object is created.  
+   **/
+  search: async (
+    context: RequestHandlerContext,	  
+    request: IEnhancedEsSearchRequest,
+    options?: ISearchOptions
+  ) => ISearchResponse<Payload=any>
+}
+```
+
+### Server Routes
+
+Each route exposes the corresponding method from the Session Service
+
+`POST /internal/session/store`
+
+`POST /internal/session/extend`
+
+`POST /internal/session/expire`
+
+`GET /internal/session/list`
```

Comment:
Nobody is expected to use the rest APIs directly though, right? I assume they just exist for the client side Session Service to call behind the scenes as an implementation detail?

Reply:
Correct, clarified in 42d99658ce5.